File: Day 1/solution.py
"""Module for calculating total distance and similarity score from input file."""


def total_distance() -> int:
    """Calculates the total distance between corresponding elements of two lists
    from "input.txt". Returns the sum of absolute differences between corresponding
    elements of the two sorted lists."""
    total_distance_number: int = 0

    left_list: list = []
    right_list: list = []

    with open("input.txt", "r", encoding="utf-8") as file:
        data: list = file.read().splitlines()
        for line in data:
            left_list.append(line.split()[0])
            right_list.append(line.split()[1])

    left_list.sort()
    right_list.sort()

    for index, left_value in enumerate(left_list):
        total_distance_number += abs(int(left_value) - int(right_list[index]))

    return total_distance_number


def similar_score() -> int:
    """Calculates the total similar score between corresponding elements of two lists from "input.txt"."""
    total_similar_score: int = 0

    left_list: list = []
    right_list: list = []

    with open("input.txt", "r", encoding="utf-8") as file:
        data: list = file.read().splitlines()
        for line in data:
            left_list.append(int(line.split()[0]))
            right_list.append(int(line.split()[1]))

    for _, left_value in enumerate(left_list):
        left_value_count = right_list.count(left_value)
        total_similar_score += left_value_count * left_value

    return total_similar_score


if __name__ == "__main__":
    print(f"This is the total distance score: {total_distance()}")
    print(f"This is the similar score: {similar_score()}")

File: Day 1/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func getLists(path string) ([]int, []int) {
	inputData, err := os.ReadFile(path)
	check(err)

	var leftList, rightList []int

	lines := strings.Split(string(inputData), "\n")

	for _, line := range lines {
		newLine := strings.Fields(line)

		leftValue, err := strconv.Atoi(newLine[0])
		check(err)

		rightValue, err := strconv.Atoi(newLine[1])
		check(err)

		rightList = append(rightList, rightValue)
		leftList = append(leftList, leftValue)
	}

	sort.Slice(leftList, func(i, j int) bool {
		return leftList[i] < leftList[j]
	})

	sort.Slice(rightList, func(i, j int) bool {
		return rightList[i] < rightList[j]
	})

	return leftList, rightList
}

func totalDistance(rightList, leftList []int) int {
	var totalDistanceScore int

	for i := 0; i < len(rightList); i++ {
		totalDistanceScore += int(math.Abs(float64(rightList[i] - leftList[i])))
	}
	return totalDistanceScore
}

func similairScore(leftList, rightList []int) int {
	var similairScore int

	count := make(map[int]int)

	for _, value := range leftList {
		count[value]++
	}

	for _, value := range rightList {
		count := count[value]
		similairScore += value * count
	}

	return similairScore
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	inputPath := "input.txt"

	leftList, rightList := getLists(inputPath)

	totalDistance := totalDistance(rightList, leftList)
	similairScore := similairScore(leftList, rightList)

	fmt.Printf("Total distance: %d\n", totalDistance)
	fmt.Printf("Similair score: %d\n", similairScore)
}
